delta/plugins/archive/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/archive/ArchiveDownload.scala

```diff
@@ -0,0 +1,189 @@
+import ch.epfl.bluebrain.nexus.delta.plugins.archive.model.ArchiveReference.{FileReference, ResourceReference}
+import ch.epfl.bluebrain.nexus.delta.plugins.archive.model.ArchiveRejection._
+import ch.epfl.bluebrain.nexus.delta.plugins.archive.model.ArchiveResourceRepresentation._
+import ch.epfl.bluebrain.nexus.delta.plugins.archive.model._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.Files
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.model.FileRejection
+import ch.epfl.bluebrain.nexus.delta.rdf.RdfError
+import ch.epfl.bluebrain.nexus.delta.rdf.implicits._
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.context.RemoteContextResolution
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.encoder.JsonLdEncoder
+import ch.epfl.bluebrain.nexus.delta.rdf.utils.JsonKeyOrdering
+import ch.epfl.bluebrain.nexus.delta.sdk.Permissions.resources
+import ch.epfl.bluebrain.nexus.delta.sdk.ReferenceExchange.ReferenceExchangeValue
+import ch.epfl.bluebrain.nexus.delta.sdk.model.acls.AclAddress
+import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sdk.model.{BaseUri, ResourceRef}
+import ch.epfl.bluebrain.nexus.delta.sdk.{Acls, AkkaSource, ReferenceExchange}
+import com.typesafe.scalalogging.Logger
+import monix.bio.{IO, UIO}
+
+/**
+  * Archive download functionality.
+  */
+trait ArchiveDownload {
+
+  /**
+    * Generates an akka [[Source]] of bytes representing the tar.
+    *
+    * @param value          the archive value
+    * @param project        the archive parent project
+    * @param ignoreNotFound do not fail when resource references are not found
+    * @param caller         the caller to be used for checking for access
+    */
+  def apply(
+      value: ArchiveValue,
+      project: ProjectRef,
+      ignoreNotFound: Boolean
+  )(implicit caller: Caller): IO[ArchiveRejection, AkkaSource]
+
+}
+
+object ArchiveDownload {
+
+  /**
+    * The default [[ArchiveDownload]] implementation.
+    *
+    * @param exchanges the collection of [[ReferenceExchange]] implementations
+    * @param acls      the acls module
+    * @param files     the files module
+    * @param sort      the configuration for sorting json keys
+    */
+  class ArchiveDownloadImpl(exchanges: Set[ReferenceExchange], acls: Acls, files: Files)(implicit
+      sort: JsonKeyOrdering,
+      baseUri: BaseUri,
+      rcr: RemoteContextResolution
+  ) extends ArchiveDownload {
+
+    implicit private val logger: Logger = Logger[ArchiveDownload]
+
+    override def apply(
+        value: ArchiveValue,
+        project: ProjectRef,
+        ignoreNotFound: Boolean
+    )(implicit caller: Caller): IO[ArchiveRejection, AkkaSource] = {
+      val referenceList = value.resources.value.toList
+      for {
+        _          <- checkResourcePermissions(referenceList, project)
+        optEntries <- referenceList.traverse {
```

Comment:
We could fetch resources/files in parallel no ?

Reply:
Creating a tar is sequential... fetching things in parallel would not speed things up significantly. And given the fact that the number of references may be high, we can generate unnecessary load on the system.